File: src/class070/C6_MaximumSubmatrix.cpp
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;

class Code06_MaximumSubmatrix {
public:
    static vector<int> getMaxMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int maxSum = INT_MIN;
        vector<int> result(4); // To store the coordinates (a, b, c, d)
        vector<int> nums(m);
        for (int up = 0; up < n; ++up) {
            fill(nums.begin(), nums.end(), 0);
            for (int down = up; down < n; ++down) {
                int pre = INT_MIN;
                for (int l = 0, r = 0; r < m; ++r) {
                    nums[r] += grid[down][r];
                    if (pre >= 0) {
                        pre += nums[r];
                    } else {
                        pre = nums[r];
                        l = r;
                    }
                    if (pre > maxSum) {
                        maxSum = pre;
                        result[0] = up;    // Top row of the submatrix
                        result[1] = l;     // Left column of the submatrix
                        result[2] = down;  // Bottom row of the submatrix
                        result[3] = r;     // Right column of the submatrix
                    }
                }
            }
        }
        return result;
    }
};

File: src/class070/C6_MaximumSubmatrix.py
class Code06_MaximumSubmatrix:
    @staticmethod
    def getMaxMatrix(grid):
        n = len(grid)
        m = len(grid[0])
        max_sum = float('-inf')
        result = [0, 0, 0, 0]  # 存储结果的坐标 (a, b, c, d)
        nums = [0] * m
        for up in range(n):
            nums = [0] * m
            for down in range(up, n):
                pre = float('-inf')
                l = 0
                for r in range(m):
                    nums[r] += grid[down][r]
                    if pre >= 0:
                        pre += nums[r]
                    else:
                        pre = nums[r]
                        l = r
                    if pre > max_sum:
                        max_sum = pre
                        result = [up, l, down, r]
        return result

